### PR TITLE
Add (Int|Rat|Bv)PosExpr to core and CFA grammar support

### LINKED ISSUE #47: Cfa DSL bad parsing around 'x+1' vs 'x + 1' in expression
In a CFA expression, 'x+1' has bad parsing, because '+1' is parsed first in INT as a token:
https://github.com/ftsrg/theta/blob/master/subprojects/cfa/src/main/antlr/CfaDsl.g4#L387

It does not have the same behaviour, if it has a space between '+' and '1'.

If we have to process everything the same way, I've found 2 solutions (untested):

Here at additiveExpr:
https://github.com/ftsrg/theta/blob/master/subprojects/cfa/src/main/antlr/CfaDsl.g4#L183

If we write 

```(firstOper=(PLUS|MINUS))? ops+=multiplicativeExpr (opers+=(PLUS | MINUS) ops+=multiplicativeExpr)*```

The parsing should be OK, but now there is a special case where there is a sign before the first token.

Another solution would be modifying at negExpr:
https://github.com/ftsrg/theta/blob/master/subprojects/cfa/src/main/antlr/CfaDsl.g4#L192

There could be a `PLUS op=negExpr` which does not negate the value, but only consumes the unneeded plus sign.

### PR DESCRIPTION
**Pull request by as3810t**
Adds (Int|Rat|Bv)PosExpr to core and CFA grammar support, as a result closes #47 .

### CODE REVIEW

**Comment by hajduakos** [comment]
@radl97 can you please double check if this works? (I did not see any concrete example in the original issue #47)

**Review by radl97 — COMMENTED** [review]
Sorry for the late answer.

Thanks, I will test it tomorrow, but seems like it would not work... The problem is that my solution that I've outlined in the original issue does not seem omplete. Sorry :/ The error is with tokenizing step, not (only?) the AST building.

```
main process proc {
init loc L0
error loc LE
final loc L1
L0 -> L1 {
i := i+1
}
}
```

The problem will be that the tokenizer (NUMBER token) will see +1 before the AST is built.
INT ::= SIGN? DIGIT+ means, that whenever + is followed by a 1, they will be a token.
So I think removing the sign would make a complete solution.
Yet again, *I've not tested my claims* (either way: whether it fixes anything, or if there is anything to fix). I will check all these tomorrow.

Also, I've found minor code style issues, see the comments.

**Comment by radl97** [comment]
Both fail:
```
[rx7@rx7 theta]$ LD_LIBRARY_PATH="$PWD/lib" java -jar subprojects/cfa-cli/build/libs/theta-cfa-cli-1.6.3-all.jar --model cfa.cfa
line 6:6 extraneous input '+1' expecting {'havoc', 'assume', 'return', ID, '}'}
```

CFA:
```
main process proc {
  init loc L0
  error loc LE
  final loc L1
  var i : int
  L0 -> L1 {
    i := i+1
  }
}
```

Proposal: remove `SIGN?` from integer token here: https://github.com/ftsrg/theta/blob/pos/subprojects/cfa/src/main/antlr/CfaDsl.g4#L497

After this change, the example runs as intended.

**Comment by radl97** [comment]
One more thought:

I would probably test whether these look valid or invalid (both with/without spaces):
Valid:
+1+5
+x+1
-1+x
etc.

Invalid:
+1 + -5 (unary mixed with binary operator)
- - 1
- + 1
+ 5 * - 1 (but could be OK)

Can you test these?

File: subprojects/core/src/main/java/hu/bme/mit/theta/core/dsl/impl/ExprCreatorVisitor.java [hunk 1]
```diff
@@ -21,18 +21,7 @@
...
-import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.Geq;
-import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.Gt;
-import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.Ite;
-import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.Leq;
-import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.Lt;
-import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.Mul;
-import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.Neg;
-import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.Neq;
-import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.Sub;
+import static hu.bme.mit.theta.core.type.abstracttype.AbstractExprs.*;
```
[the remark this answers is not in the document]
Comment (as3810t) [reply subprojects/core/src/main/java/hu/bme/mit/theta/core/dsl/impl/ExprCreatorVisitor.java]: Fixed by 827a984

File: subprojects/core/src/main/java/hu/bme/mit/theta/core/utils/ExprSimplifier.java [hunk 2]
```diff
@@ -51,19 +51,7 @@
...
-import hu.bme.mit.theta.core.type.rattype.RatGtExpr;
-import hu.bme.mit.theta.core.type.rattype.RatLeqExpr;
-import hu.bme.mit.theta.core.type.rattype.RatLitExpr;
-import hu.bme.mit.theta.core.type.rattype.RatLtExpr;
-import hu.bme.mit.theta.core.type.rattype.RatMulExpr;
-import hu.bme.mit.theta.core.type.rattype.RatNegExpr;
-import hu.bme.mit.theta.core.type.rattype.RatNeqExpr;
-import hu.bme.mit.theta.core.type.rattype.RatSubExpr;
-import hu.bme.mit.theta.core.type.rattype.RatType;
+import hu.bme.mit.theta.core.type.rattype.*;
```
[the remark this answers is not in the document]
Comment (as3810t) [reply subprojects/core/src/main/java/hu/bme/mit/theta/core/utils/ExprSimplifier.java]: Fixed by 827a984

File: subprojects/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3ExprTransformer.java [hunk 3]
```diff
@@ -55,18 +55,7 @@
...
-import hu.bme.mit.theta.core.type.rattype.RatGeqExpr;
-import hu.bme.mit.theta.core.type.rattype.RatGtExpr;
-import hu.bme.mit.theta.core.type.rattype.RatLeqExpr;
-import hu.bme.mit.theta.core.type.rattype.RatLitExpr;
-import hu.bme.mit.theta.core.type.rattype.RatLtExpr;
-import hu.bme.mit.theta.core.type.rattype.RatMulExpr;
-import hu.bme.mit.theta.core.type.rattype.RatNegExpr;
-import hu.bme.mit.theta.core.type.rattype.RatNeqExpr;
-import hu.bme.mit.theta.core.type.rattype.RatSubExpr;
+import hu.bme.mit.theta.core.type.rattype.*;
```
[the remark this answers is not in the document]
Comment (as3810t) [reply subprojects/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3ExprTransformer.java]: Fixed by 827a984

**Comment by as3810t** [comment]
@radl97 With the proposed fix, your examples are parsed as following:

```
+1+5 -> (+1) + 5
+x+1 -> (+x) + 1
-1+x -> (-1) + x

+1 + -5 -> (+1) + (-5)
- - 1 -> -(-(1))
- + 1 -> -(+(1))
+ 5 * - 1 -> (+5) * (-1)
```

**Comment by radl97** [comment]
Well, it's a bit permissive, but ultimately it's not a problem.

Looks good to me

**Comment by hajduakos** [comment]
Looks good. In the end, the CFA DSL is not really meant to be written by humans, but rather produced automatically by frontends. These frontends usually put brackets everywhere to make sure.

**Pull request merged**